### ISSUE TITLE
"# of Slow Layers" Doesn't Function Correctly

### ISSUE THREAD

**Issue by CCS86** [SoftFever/OrcaSlicer#4898]
### Operating System (OS) Windows 10 x64

The "number of slow layers" setting doesn't appear to function correctly. Setting it to "1" or "2" appears to have no effect at all, as layer 2 is full speed. Setting it to "3" finally has an effect on layer 2, but only slows the outer wall.

I expect that when it is set to "0", we get a first layer at the first layer speed, and then straight to commanded speeds on layer 2.

When it is set to "1", then layer 2 would be slowed down, and full speed on layer 3.

When it is set to "2", then layers 2 and 3 will be slowed, and full speed on layer 4.

And so on.

I also expect that all line types would be slowed by this modifier.

![image](https://github.com/SoftFever/OrcaSlicer/assets/12439543/78a5bd0c-c298-4772-9cbb-8f7bb71ad849)

![image](https://github.com/SoftFever/OrcaSlicer/assets/12439543/5d5829b0-074d-49d7-ac58-1bbe884b55af)

![image](https://github.com/SoftFever/OrcaSlicer/assets/12439543/0f4f4366-34e4-48af-a222-e702d7f5072d)

![image](https://github.com/SoftFever/OrcaSlicer/assets/12439543/00487ac4-92c2-43d6-ba33-e7390c16ffd7)

![image](https://github.com/SoftFever/OrcaSlicer/assets/12439543/6f9fdf9a-ddf0-4c72-8b05-31afd11ba45e)

**Comment by bma73-bma** [comment]
It doesn't matter what number I add in the "Number of slow layers", all walls and the infill on the first layer is always printed with the speed I put in "First layer" and "First layer infill", and from layer 2 and on it's full speed.

Logical to me would be a checkbox for Enable/Disable the section "First layer speed". If I enable the section the default value in "Numbers of slow layers" should be 1 and the first layer should be printed slow. If I put 2 there the first 2 layers should be printed slow and so on.

Alternative to a checkbox would be if I put 0 as the number of slow layers it means full speed from start.

**Comment by CCS86** [comment]
@SoftFever, can you reproduce?

**Comment by mtc1122** [comment]
Confirmed that it only slows first layer when I've been using the setting recently as well.

**Comment by matheusjat** [comment]
![Captura de tela 2024-06-01 111007](https://github.com/SoftFever/OrcaSlicer/assets/15360253/af79f390-80e1-4b9a-946d-3458dcd5b9a5)

with high values, the function starts to work

**Comment by CCS86** [comment]
Bump

**Comment by WimVDK** [comment]
still happening in 2.2.0. Slow layer numbers is ignored with values like 3 slow layers

**Comment by JoshMcguigan** [comment]
I'm also seeing this on `2.2.0`.

0 slow layers, slows down first layer:

<img width="1440" alt="image" src="https://github.com/user-attachments/assets/6808498e-a48d-498a-b0d7-1d6a99a22516">

10 slow layers, seems the same/similar as 0:

<img width="1440" alt="image" src="https://github.com/user-attachments/assets/a4edd09e-a7ec-4468-a26d-b6061aee9742">

50 slow layers, first 5 or so layers are slowed down:

<img width="1440" alt="image" src="https://github.com/user-attachments/assets/ecd61180-b293-45df-b489-8eb71218c94a">

100 slow layers, first 10 or so layers are slowed down:

<img width="1440" alt="image" src="https://github.com/user-attachments/assets/33127232-c221-412e-b59a-630544c4ea1d">

**Comment by FrankHartog** [comment]
I also noticed that "Number of slow layers" doesn't seem to have any effect. Only the first layer is slowed down, the following layers are printed at full speed.

**Comment by nicodoettling** [comment]
This seems to be normal behaviour. For the slow layers Orca Slicer does not set the speed to the first layer speed, but interpolates between first layer speed and normal speed.

You can see this in line 5266 of GCode.cpp: https://github.com/SoftFever/OrcaSlicer/blob/5d2e7e4dc42f4b510e19dccb2159342acf42fb4a/src/libslic3r/GCode.cpp#L5265

                speed = std::min(
                    speed,
                    Slic3r::lerp(first_layer_speed, speed,
                                 (double)_layer / m_config.slow_down_layers));

if you comment this out and replace it with 

speed = std::min(speed,first_layer_speed);

and recompile you get the desired behaviour, i.e. the slow layers are as slow as the first one.

**Comment by WimVDK** [comment]
That's designed behavior, not normal. There's a difference. With more and more high speed printers, the code may need adjusting. First layers at 80-100mm/s for 'slow' speed isn't slow at all. That's almost twice the top speed many 3d printers were capable of only a few years ago

**Comment by CCS86** [comment]
> This seems to be normal behaviour. For the slow layers Orca Slicer does not set the speed to the first layer speed, but interpolates between first layer speed and normal speed.


It's not working correctly. 

I showed that it was only slowing down the outer walls, not the whole layer like it should. 

It also has to be set much higher than it should to have an effect.

**Comment by nicodoettling** [comment]
I did a few test prints and found that the issue comes more down to acceleration rather than speed. If you set the acceleration to the of all layers to that of the first layer, do you still see the issue?

On that note, I modified the code to also set the acceleration of the slow layers to that of the first layer, and the result is what you would expect, the slow layers behave exactly like the first layer.

**Comment by bma73-bma** [comment]
When I looked at Creality Print the description of number of slow layers says that it's the number of layers over which the speed is gradually increased. So if that's the case it's not supposed to print at the slow speed on all the layers specified in this setting, it will ramp up the speed. So of course if I tell it to print slow for 3 layers and first layer speed is 50 and the other layer speed is 300 it will look like a huge increase in speed between layer 1 and 2.

**Comment by CCS86** [comment]
> I did a few test prints and found that the issue comes more down to acceleration rather than speed. If you set the acceleration to the of all layers to that of the first layer, do you still see the issue?
> 
> On that note, I modified the code to also set the acceleration of the slow layers to that of the first layer, and the result is what you would expect, the slow layers behave exactly like the first layer.

This has nothing to do with acceleration. We are talking about print speed.

**Comment by CCS86** [comment]
> When I looked at Creality Print the description of number of slow layers says that it's the number of layers over which the speed is gradually increased. So if that's the case it's not supposed to print at the slow speed on all the layers specified in this setting, it will ramp up the speed. So of course if I tell it to print slow for 3 layers and first layer speed is 50 and the other layer speed is 300 it will look like a huge increase in speed between layer 1 and 2.

Yes, obviously the fewer the "slow layers" and the faster the normal print speed, the larger the jump in speed between layers. 

Look at the photos I posted again. The first 2 have the setting disabled (slow layers = 0), and it shows the speeds for layer 1 and layer 2, without any modification. 

The rest of the photos stay on layer number 2. Slower layers gets set to 1, and layer 2's speed doesn't change. It gets set to 2, and it still doesn't change. It gets set to 3 and only the outer wall slows down.

**Comment by WimVDK** [comment]
There is no example. Set a number of slow layers in orca slicer, print. Watch orcaslicer ignore them. There is no visible slowness unless you set the number absurdly high like 30 layers

**Comment by CCS86** [comment]
I just tested it in the latest nightly, and it has been fixed. I'm not sure what version applied the fix, but it is working now:

![Image](https://github.com/user-attachments/assets/07828ee9-b8fa-42f2-8307-07b131f71aab)
